Added overview of C# Boolean logical operators (#11656)
dotnet/docs: docs/csharp/language-reference/operators/boolean-logical-operators.md

  - "^_CSharpKeyword"
  - "|_CSharpKeyword"
  - "&&_CSharpKeyword"
  - "||_CSharpKeyword"
helpviewer_keywords: 
  - "logical operators [C#]"
  - "short-circuiting operators [C#]"
  - "logical negation operator [C#]"
  - "NOT operator [C#]"
  - "! operator [C#]"
  - "AND operator [C#]"
  - "ampersand operator [C#]"
  - "conjunction operator [C#]"
  - "& operator [C#]"
  - "exclusive OR operator [C#]"
  - "XOR operator [C#]"
  - "^ operator [C#]"
  - "OR operator [C#]"
  - "disjunction operator [C#]"
  - "| operator [C#]"
  - "conditional logical AND operator [C#]"
  - "short-circuiting logical AND operator [C#]"
  - "&& operator [C#]"
  - "conditional logical OR operator [C#]"
  - "short-circuiting logical OR operator [C#]"
  - "|| operator [C#]"
---
# Boolean logical operators (C# Reference)

The following operators perform logical operations with [bool](../keywords/bool.md) operands:

- Unary [`!` (logical negation)](#logical-negation-operator-) operator.
- Binary [`&` (logical AND)](#logical-and-operator-), [`|` (logical OR)](#logical-or-operator-), and [`^` (logical exclusive OR)](#logical-exclusive-or-operator-) operators. Those operators always evaluate both operands.

> - Binary [`&&` (conditional logical AND)](#conditional-logical-and-operator-) and [`||` (conditional logical OR)](#conditional-logical-or-operator-) operators. Those operators evaluate the second operand only if it's necessary.

Author: Given that listed operators are for `bool` operands, I haven't included `true` and `false` operators here; also, I assume their usage is very rare with comparison to the operators in this article. Anyway, the `true` and `false` operators are mentioned in the "Operator overloadability" section.

Please let me know, if it's fine.

Reviewer: I agree. The `true` and `false` operators should be a separate topic.

The `&` and `|` operators are also defined for `bool?` operands. For more information, see [The bool? type](../../programming-guide/nullable-types/using-nullable-types.md#the-bool-type) section of the [Using nullable types](../../programming-guide/nullable-types/using-nullable-types.md) article.

> For the operands of [integral](../keywords/integral-types-table.md) types, the `&`, `|`, and `^` operators perform bitwise logical operations.

Author: No links to operators reference by design (though individual sections below have links): I'll add the link to the overview of bitwise operators here. To create such an overview is my next PR.

Reviewer: I think that's a good plan.

I am watching traffic patterns to determine when readers land in the wrong article, and which article they are trying to find.

> ## Logical negation operator !

Author: Should I mention the null-forgiving operator and add the link to the nullable reference types overview? Or that is to be done after the official release of C# 8.0?

Reviewer: Let's wait on that. I'd rather add a link to a page on the null forgiveness operator (which doesn't exist yet).

A nice side-effect of using the displayName in the TOCs is that it's easy to find which pages discuss an operator and a different usage. 😸

The unary `!` operator computes logical negation of its operand. That is, it produces `true`, if the operand evaluates to `false`, and `false`, if the operand evaluates to `true`:

[!code-csharp-interactive[logical negation](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#Negation)]

> ## Logical AND operator &amp;

Author: @BillWagner could you please tell me what anchor is generated for this header on the review site. I use `logical-and-operator-` in this article (does it work?); however, Visual Studio Code doesn't recognize such a link and doesn't navigate me to the section.

Reviewer: The anchor in the review site is `logical-and-operator-` .  Good guess!

Author: Cool, thanks!

The `&` operator computes the logical AND of its operands. The result of `x & y` is `true` if both `x` and `y` evaluate to `true`. Otherwise, the result is `false`.

> The `&` operator evaluates both operands even if the first operand evaluates to `false`, so that the result must be `false` regardless of the value of the second operand. The following example demonstrates that behavior:

Reviewer: As you introduce the concept of short-circuiting (or not), I think it's important to explain that `x` and `y` could be Boolean expressions (including method calls). In other words, the operands are not limited to variables.

Author: @BillWagner hm, I had a feeling I've explained it: the "evaluates" word means that an operand could be an expression (or not?). Also, at the example one of operands is a local function call, clearly not a variable. Do you suggest explicitly state in the article that operands might be expressions? I feel a bit odd here, because operands typically could be expressions and only exceptions to that rule are usually mentioned in the docs (for example, the `++` and `--` operators).

Author: @BillWagner I've updated the introductions to the examples for the `&`, `|`, `&&`, and `||` operators. Also, examples are updated: dotnet/samples#797

What do you think of a new text?

> The `&` operator evaluates both operands even if the first operand evaluates to `false`, so that the result must be `false` regardless of the value of the second operand. The following example demonstrates that behavior:

Reviewer: This is a super-nit, but it's more accurate to say it as follows:

> ...even if the first operand doesn't evaluate to `true`, so that …

That difference only matters for types that override operator `true` and operator `false` *and* could be in a state where both operators return `false`.

The pertinent clause in the spec is [here](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#conditional-logical-operators). See the two bullets following the sample.

Author: @BillWagner I like this nit-picking; also your comment confused me a bit.

According to the spec `x && y` is evaluated as `T.false(x) ? x : T.&(x, y)`. 

First, the `true` operator doesn't participate at all in definition and we cannot refer to it then. I assume when you say
>`x` doesn't evaluate to `true`

you talk about the `true` operator and mean that "`x` is not definitely true", or `T.true(x)` returns `false`. Do I understand you correctly?

Second, back to `x && y` as `T.false(x) ? x : T.&(x, y)`. I read that definition as 
> If the first operand is definitely false, do short-circuit. Otherwise, call the `&` operator (which means the second operator is evaluated as well).

So, when we say that the second operand is evaluated always, we emphasize that "always" means even if the first operand is definitely false (so even if there is condition for short-circuiting). I think that "the first operand is definitely false" can be rephrased as "the first operand evaluates to `false`".

If there is a flaw in my reasoning, please let me know. I remember I've looked at that spec and tried to build a sentence with keeping user-defined conditional operators in mind, but might have missed or misinterpreted something.

Reviewer: You're analysis is correct. I'll resolve this conversation.

I had started editing and suggesting, then pulled out my copy of the spec as I got farther along. I was trying to make small edits without checking. That's always a mistake.

[!code-csharp-interactive[logical AND](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#And)]

The [conditional logical AND operator](#conditional-logical-and-operator-) `&&` also computes the logical AND of its operands, but evaluates the second operand only if the first operand evaluates to `true`.

For the operands of integral types, the `&` operator computes [bitwise logical AND](and-operator.md#integer-logical-bitwise-and-operator) of its operands. The unary `&` operator is the [address-of operator](and-operator.md#unary-address-of-operator).

> ## Logical exclusive OR operator ^

Author: I order operator sections in each overview by the operator precedence.

The `^` operator computes the logical exclusive OR, also known as the logical XOR, of its operands. The result of `x ^ y` is `true` if `x` evaluates to `true` and `y` evaluates to `false`, or `x` evaluates to `false` and `y` evaluates to `true`. Otherwise, the result is `false`. That is, for Boolean operands, the `^` operator computes the same result as the [inequality operator](equality-operators.md#inequality-operator-) `!=`.

[!code-csharp-interactive[logical exclusive-OR](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#Xor)]

For the operands of integral types, the `^` operator computes [bitwise logical exclusive OR](xor-operator.md) of its operands.

## Logical OR operator |

The `|` operator computes the logical OR of its operands. The result of `x | y` is `true` if either `x` or `y` evaluates to `true`. Otherwise, the result is `false`.

> The `|` operator evaluates both operands even if the first operand evaluates to `true`, so that the result must be `true` regardless of the value of the second operand. The following example demonstrates that behavior:

Reviewer: This should be updated similarly to my previous comment, except that the first operand doesn't evaluate to `false`.

Author: Please check my comment to the `&` operator section.

Reviewer: Same answer... Resolving.

[!code-csharp-interactive[logical OR](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#Or)]

> The [conditional logical OR operator](#conditional-logical-or-operator-) `||` also computes the logical OR of its operands, but evaluates the second operand only if the first operand evaluates to `false`.

Reviewer: Again:  

> only if the first operand doesn't evaluate to `true`.

Author: @BillWagner while your suggestion is correct, I'm afraid readers would translate it to "...evaluates to `false`" anyway ("doesn't evaluate to `true`" might be hard to comprehend), with most proactive of them submitting PR to "fix" wording.

I suggest another fix. I've noticed that I'm not consistent when talking about short-circuiting behavior. Sometimes I mention condition for short-circuiting ("...the first operand evaluates to `true`" in case of `||`), in other cases — condition for evaluation of the second operand ("...the first operand doesn't evaluate to `true`"). I think only one condition should be mentioned. I suggest to describe condition for short-circuit; in particular, because the [spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#user-defined-conditional-logical-operators) definition of user-defined conditional operators says when to short-circuit.

Then, the paragraph in question would read:
>The conditional logical OR operator `||` also computes the logical OR of its operands, but doesn't evaluate the second operand if the first operand evaluates to `true`.

Reviewer: Resolved again.

For the operands of integral types, the `|` operator computes [bitwise logical OR](or-operator.md) of its operands.

> ## Conditional logical AND operator &amp;&amp;

Author: Also, what is the generated anchor for this header? Does `conditional-logical-and-operator-` work?

Reviewer: This one is `conditional-logical-and-operator-`.  Good guess again.

> The conditional logical AND operator `&&`, also known as the "short-circuiting" logical AND operator, computes the logical AND of its operands. The result of `x && y` is `true` if both `x` and `y` evaluate to `true`. Otherwise, the result is `false`. If the first operand evaluates to `false`, the second operand is not evaluated and the result of operation is `false`. The following example demonstrates that behavior:

Reviewer: ```suggestion
The conditional logical AND operator `&&`, also known as the "short-circuiting" logical AND operator, computes the logical AND of its operands. The result of `x && y` is `true` if both `x` and `y` evaluate to `true`. Otherwise, the result is `false`. The second operand is evaluated only when the first operand is not `false`. The following example demonstrates that behavior:
```

Author: @BillWagner is it fine to replace in your suggestion "first operand is not `false`" with "first operand doesn't evaluate to `false`" I understand that it cannot be replaced with "first operand evaluates to `true`"

Reviewer: I like your wording better. Let's go with it.

[!code-csharp-interactive[conditional logical AND](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#ConditionalAnd)]

The [logical AND operator](#logical-and-operator-) `&` also computes the logical AND of its operands, but always evaluates both operands.

## Conditional logical OR operator ||

> The conditional logical OR operator `||`, also known as the "short-circuiting" logical OR operator, computes the logical OR of its operands. The result of `x || y` is `true` if either `x` or `y` evaluates to `true`. Otherwise, the result is `false`. If the first operand evaluates to `true`, the second operand is not evaluated and the result of operation is `true`. The following example demonstrates that behavior:

Reviewer: ```suggestion
The conditional logical OR operator `||`, also known as the "short-circuiting" logical OR operator, computes the logical OR of its operands. The result of `x || y` is `true` if either `x` or `y` evaluates to `true`. Otherwise, the result is `false`. The second operand is evaluated only when the first operand is not `true`. The following example demonstrates that behavior:
```

[!code-csharp-interactive[compound assignment](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#CompoundAssignment)]

The conditional logical operators `&&` and `||` don't support compound assignment.

## Operator precedence

The following list orders logical operators starting from the highest precedence to the lowest:

- Logical negation operator `!`
- Logical AND operator `&`
- Logical exclusive OR operator `^`
- Logical OR operator `|`
- Conditional logical AND operator `&&`
- Conditional logical OR operator `||`

Use parentheses, `()`, to change the order of evaluation imposed by operator precedence:

[!code-csharp-interactive[operator precedence](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#Precedence)]

For the complete list of C# operators ordered by precedence level, see [C# operators](index.md).

## Operator overloadability

A user-defined type can [overload](../keywords/operator.md) the `!`, `&`, `|`, and `^` operators. When a binary operator is overloaded, the corresponding compound assignment operator is also implicitly overloaded. A user-defined type cannot explicitly overload a compound assignment operator.

> A user-defined type cannot overload the conditional logical operators `&&` and `||`. However, if a user-defined type overloads the `&` operator (the `|` operator) and the [true and false operators](../keywords/true-false-operators.md) in a certain way, the `&&` operation (the `||` operation) can be evaluated for the operands of that type. For more information, see the [User-defined conditional logical operators](~/_csharplang/spec/expressions.md#user-defined-conditional-logical-operators) section of the [C# language specification](~/_csharplang/spec/introduction.md).

Reviewer: The parenthetical expressions "(the `|` operator)" and "(the `||` operator) look odd to me. I don't have a good suggestion right now. I'll keep noodling. But, if you come up with a better wording make an edit here.

Author: I like you've mentioned it. I'm also not fully satisfied with parentheses. Wander more about it, then.

Author: Rephrased.
